sdk/database/newdbplugin/grpc_server.go

```diff
@@ -0,0 +1,180 @@
+package newdbplugin
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/hashicorp/vault/sdk/database/newdbplugin/proto"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var _ proto.DatabaseServer = gRPCServer{}
+
+type gRPCServer struct {
+	impl Database
+}
+
+// Initialize the database plugin
+func (g gRPCServer) Initialize(ctx context.Context, request *proto.InitializeRequest) (*proto.InitializeResponse, error) {
+	rawConfig := structToMap(request.ConfigData)
+
+	dbReq := InitializeRequest{
+		Config:           rawConfig,
+		VerifyConnection: request.VerifyConnection,
+	}
+
+	dbResp, err := g.impl.Initialize(ctx, dbReq)
+	if err != nil {
+		return &proto.InitializeResponse{}, status.Errorf(codes.Internal, "failed to initialize: %s", err)
+	}
+
+	newConfig, err := mapToStruct(dbResp.Config)
+	if err != nil {
+		return &proto.InitializeResponse{}, status.Errorf(codes.Internal, "failed to marshal new config to JSON: %s", err)
+	}
+
+	resp := &proto.InitializeResponse{
+		ConfigData: newConfig,
+	}
+
+	return resp, nil
+}
+
+func (g gRPCServer) NewUser(ctx context.Context, req *proto.NewUserRequest) (*proto.NewUserResponse, error) {
+	if req.GetUsernameConfig() == nil {
+		return &proto.NewUserResponse{}, status.Errorf(codes.InvalidArgument, "missing username config")
+	}
+
+	var expiration time.Time
+
+	if req.GetExpiration() != nil {
+		exp, err := ptypes.Timestamp(req.GetExpiration())
+		if err != nil {
+			return &proto.NewUserResponse{}, status.Errorf(codes.InvalidArgument, "unable to parse expiration date: %s", err)
+		}
+		expiration = exp
+	}
+
+	dbReq := NewUserRequest{
+		UsernameConfig: UsernameMetadata{
+			DisplayName: req.GetUsernameConfig().GetDisplayName(),
+			RoleName:    req.GetUsernameConfig().GetRoleName(),
+		},
+		Statements: Statements{
+			Commands: req.GetStatements().GetCommands(),
+		},
+		Password:   req.GetPassword(),
+		Expiration: expiration,
+	}
+
+	dbResp, err := g.impl.NewUser(ctx, dbReq)
+	if err != nil {
+		return &proto.NewUserResponse{}, status.Errorf(codes.Internal, "unable to create new user: %s", err)
+	}
+
+	resp := &proto.NewUserResponse{
+		Username: dbResp.Username,
+	}
+	return resp, nil
+}
+
+func (g gRPCServer) UpdateUser(ctx context.Context, req *proto.UpdateUserRequest) (*proto.UpdateUserResponse, error) {
+	if req.GetUsername() == "" {
+		return &proto.UpdateUserResponse{}, status.Errorf(codes.InvalidArgument, "no username provided")
+	}
+
+	dbReq, err := getUpdateUserRequest(req)
+	if err != nil {
+		return &proto.UpdateUserResponse{}, status.Errorf(codes.InvalidArgument, err.Error())
```

Comment:
Do we want to return `codes.InvalidArgument` for a call to `UpdateUser()` that doesn't change fields of the user? For example if `!hasChange(dbReq)` is true on line 131.

If this results in an eventual http response with a 400 status code to clients, we may want to consider if that's appropriate (usually malformed syntax, etc).

Reply:
I'm thinking we should do whatever is consistent with other Vault APIs for updating a resource when nothing is actually being changed. I'm also not sure how these error codes map to the eventual http response status code that a client would receive. Something to consider :)

Reply:
I think it's appropriate to require a change when requesting to update a user. Since this isn't an "open" endpoint that can be called by anyone but should only be called by Vault itself, I see this as a notification that something in Vault went wrong.

It doesn't result in an HTTP status code since this is a gRPC code, however it is the gRPC equivalent of a 400 code.

Reply:
That makes sense. I was thinking along the line that this could _eventually_ result in a 400 being returned to an actual Vault client, but that depends on how this gRPC response error is handled as you've mentioned.

No issue. Resolving.